org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/TabbedDesktop.java

```diff
@@ -87,6 +88,7 @@ public ProcessDialog openProcessDialog(int processId, boolean soTrx) {
      */
 	public ADForm openForm(int formId) {
 		ADForm form = ADForm.openForm(formId);
+		Env.setPredefinedVariables(Env.getCtx(), form.getWindowNo(), getPredefinedContextVariables());
```

Comment:
shouldn't this happens before the openForm method and the form implementation class can get access to those variables in the openForm method ?

Reply:
I tried that, but not possible, we don't have the windowNo before that

Reply:
perhaps we should just pass the context variables into the openform method ( and do the Env.setPredefinedVariables(Env.getCtx(), form.getWindowNo(), getPredefinedContextVariables()); in it ) ? I think it should be more straightforward to process the context variables in the form's initForm method.

Reply:
Tried that [FailedPatch4713.txt](https://github.com/idempiere/idempiere/files/6021603/FailedPatch4713.txt)

Still is not possible, the form is open in Extensions.getForm - and to pass the parameters there we would need to change the IFormFactory to receive the parameter.

Reply:
We don't have to change Extensions.getForm, we can add "Env.setPredefinedVariables(Env.getCtx(), form.getWindowNo(), predefinedContextVariables)"; prior to "form.init(adFormID, name);" (around line 180) and the context variables would be available when the ADForm.initForm() method is invoke.

Reply:
OK, I'm doing the suggested change.

The problem was that I was testing with WResetPassword.

So, this approach will work fine for classes extending ADForm, but not for the forms that are developed implementing IFormController, those will require anyways something like the onPageAttached mechanism, but it works fine too, I tested it.

Reply:
It can works with WResetPassword too. Instead of form = new CustomForm(), you use form = new CustomForm() { protected void initForm() { ... } };